File: Python/Leetcode Easy/111MinimumDepthOfBinaryTree.py
#https://leetcode.com/problems/minimum-depth-of-binary-tree/description/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root == None: return 0
        queue = deque([(root, 1)])
        while True:
            node, level = queue.popleft()
            if node == None: continue
            if node.left == None and node.right == None: return level
            queue.append((node.left, level+1))
            queue.append((node.right, level+1))
        

File: C++/Leetcode Easy/111MinimumDepthOfBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <list>
using namespace std;
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (!root) return 0;
        int res = 0;
        list<TreeNode*> queue = {root};
        while (!queue.empty()) {
            int n = queue.size();
            ++res;
            for (int i = 0; i < n; ++i) {
                TreeNode* curr = queue.front();
                queue.pop_front();
                if (!curr->left && !curr->right) return res;
                if (curr->left) queue.push_back(curr->left);
                if (curr->right) queue.push_back(curr->right);
            }
        }
        return -1;
    }
};